File: src/main/java/de/whitefrog/neobase/model/rest/QueryField.java
package de.whitefrog.neobase.model.rest;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

import java.util.Arrays;

/**
* Created by jochen on 12/01/16.
*/
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY,
  getterVisibility = JsonAutoDetect.Visibility.NONE,
  setterVisibility = JsonAutoDetect.Visibility.NONE)
public class QueryField {
  private String field;
  @JsonInclude(JsonInclude.Include.NON_DEFAULT)
  private int skip = 0;
  @JsonInclude(JsonInclude.Include.NON_DEFAULT)
  private int limit = SearchParameter.DefaultLimit;
  @JsonInclude(JsonInclude.Include.NON_EMPTY)
  private FieldList subFields = new FieldList();

  public QueryField() {}
  public QueryField(String field) {
    this(field, false);
  }
  public QueryField(String field, boolean addAll) {
    if(field.contains(".")) {
      String[] fields = field.split("\\.", 2);
      parseField(fields[0]);
      this.subFields = FieldList.parseFields(Arrays.asList(fields[1]), addAll);
    } else {
      parseField(field);
    }
  }
  
  private void parseField(String field) {
    if(field.contains("(")) {
      this.field = field.substring(0, field.indexOf("("));
      String limit = field.substring(field.indexOf("(") + 1, field.length() - 1);
      if(limit.contains(";")) {
        this.skip = Integer.parseInt(limit.substring(0, limit.indexOf(";")));
        limit = limit.substring(limit.indexOf(";") + 1);
        this.limit = limit.equals("max")? Integer.MAX_VALUE: Integer.parseInt(limit);
      } else {
        this.limit = limit.equals("max")? Integer.MAX_VALUE: Integer.parseInt(limit);
      }
    } else {
      this.field = field;
    }
  }

  public String field() {
    return field;
  }

  public int limit() {
    return limit;
  }

  public void limit(int limit) {
    this.limit = limit;
  }
  
  public int skip() { return skip; }
  public void skip(int skip) { this.skip = skip; }

  public FieldList subFields() {
    return subFields;
  }

  public void subFields(FieldList fields) {
    this.subFields = fields;
  }
  
  public void subFields(QueryField... fd) {
    this.subFields.addAll(Arrays.asList(fd));
  }

  @Override
  public int hashCode() {
    return new HashCodeBuilder()
      .append(field)
      .toHashCode();
  }

  @Override
  public boolean equals(Object obj) {
    if(!(obj instanceof QueryField)) return false;
    QueryField other = (QueryField) obj;
    return new EqualsBuilder()
      .append(field, other.field)
      .isEquals();
  }

  @Override
  public String toString() {
    return field + (subFields().isEmpty()? "": " (" + subFields.size() + " subfields)");
  }
}

File: src/main/kotlin/de/whitefrog/neobase/model/rest/QueryField.kt
package de.whitefrog.neobase.model.rest

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.JsonInclude
import org.apache.commons.lang.builder.EqualsBuilder
import org.apache.commons.lang.builder.HashCodeBuilder

import java.util.Arrays

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE, setterVisibility = JsonAutoDetect.Visibility.NONE)
class QueryField @JvmOverloads constructor(field: String, addAll: Boolean = false) {
  private var field: String? = null
  @JsonInclude(JsonInclude.Include.NON_DEFAULT)
  private var skip = 0
  @JsonInclude(JsonInclude.Include.NON_DEFAULT)
  private var limit = SearchParameter.DefaultLimit
  @JsonInclude(JsonInclude.Include.NON_EMPTY)
  private var subFields = FieldList()

  init {
    if (field.contains(".")) {
      val fields = field.split("\\.".toRegex(), 2).toTypedArray()
      parseField(fields[0])
      this.subFields = FieldList.parseFields(Arrays.asList(fields[1]), addAll)
    } else {
      parseField(field)
    }
  }

  private fun parseField(field: String) {
    if (field.contains("(")) {
      this.field = field.substring(0, field.indexOf("("))
      var limit = field.substring(field.indexOf("(") + 1, field.length - 1)
      if (limit.contains(";")) {
        this.skip = Integer.parseInt(limit.substring(0, limit.indexOf(";")))
        limit = limit.substring(limit.indexOf(";") + 1)
        this.limit = if (limit == "max") Integer.MAX_VALUE else Integer.parseInt(limit)
      } else {
        this.limit = if (limit == "max") Integer.MAX_VALUE else Integer.parseInt(limit)
      }
    } else {
      this.field = field
    }
  }

  fun field(): String {
    return field ?: ""
  }

  fun limit(): Int {
    return limit
  }

  fun limit(limit: Int) {
    this.limit = limit
  }

  fun skip(): Int {
    return skip
  }

  fun skip(skip: Int) {
    this.skip = skip
  }

  fun subFields(): FieldList {
    return subFields
  }

  fun subFields(fields: FieldList) {
    this.subFields = fields
  }

  fun subFields(vararg fd: QueryField) {
    this.subFields.addAll(Arrays.asList(*fd))
  }

  override fun hashCode(): Int {
    return HashCodeBuilder()
      .append(field)
      .toHashCode()
  }

  override fun equals(other: Any?): Boolean {
    if (other !is QueryField) return false
    return EqualsBuilder()
      .append(field, other.field)
      .isEquals
  }

  override fun toString(): String {
    return field!! + if (subFields().isEmpty()) "" else " (" + subFields.size + " subfields)"
  }
}
